common/CommonAPI.h

```diff
@@ -722,23 +711,281 @@ class CommonAPI
+
+	nbl::video::IGPUImage* getTripleBufferTarget(
+		uint32_t frameIx, uint32_t w, uint32_t h, 
+		nbl::asset::E_FORMAT surfaceFormat, 
+		nbl::core::bitflag<nbl::asset::IImage::E_USAGE_FLAGS> imageUsageFlags)
+	{
+		uint32_t bufferIx = frameIx % 2;
+		auto& image = m_tripleBufferRenderTargets.begin()[bufferIx];
+
+		if (!image || image->getCreationParameters().extent.width < w || image->getCreationParameters().extent.height < h)
+		{
+			auto logicalDevice = getLogicalDevice();
+			nbl::video::IGPUImage::SCreationParams creationParams;
+			creationParams.type = nbl::asset::IImage::ET_2D;
+			creationParams.samples = nbl::asset::IImage::ESCF_1_BIT;
+			creationParams.format = surfaceFormat;
+			creationParams.extent = { w, h, 1 };
+			creationParams.mipLevels = 1;
+			creationParams.arrayLayers = 1;
+			creationParams.usage = imageUsageFlags | nbl::asset::IImage::EUF_TRANSFER_SRC_BIT;
+
+			image = logicalDevice->createImage(std::move(creationParams));
+			auto memReqs = image->getMemoryReqs();
+			memReqs.memoryTypeBits &= logicalDevice->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+			logicalDevice->allocate(memReqs, image.get());
+
+			onCreateResourcesWithTripleBufferTarget(image, bufferIx);
+		}
+
+		return image.get();
+	}
+#endif
 public:
 	GraphicalApplication(
 		const std::filesystem::path& _localInputCWD,
 		const std::filesystem::path& _localOutputCWD,
 		const std::filesystem::path& _sharedInputCWD,
 		const std::filesystem::path& _sharedOutputCWD
-	) : nbl::system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
-	void recreateSurface() override
+	) : nbl::system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD),
+		CommonAPI::CommonAPIEventCallback(nullptr, nullptr),
+		m_qRetiredSwapchainResources(),
+		m_imageSwapchainIterations{}
+	{}
+
+	std::unique_lock<std::mutex> recreateSwapchain(
+		uint32_t w, uint32_t h, 
+		nbl::video::ISwapchain::SCreationParams& swapchainCreationParams, 
+		nbl::core::smart_refctd_ptr<nbl::video::ISwapchain>& swapchainRef)
 	{
+		auto logicalDevice = getLogicalDevice();
+		std::unique_lock guard(m_swapchainPtrMutex);
+		CommonAPI::createSwapchain(
+			nbl::core::smart_refctd_ptr<nbl::video::ILogicalDevice>(logicalDevice),
+			swapchainCreationParams, 
+			w, h, 
+			swapchainRef);
+		assert(swapchainRef);
+		m_swapchainIteration++;
+
+		return guard;
 	}
-	void onResize(uint32_t w, uint32_t h) override
+
+	void waitForFrame(
+		uint32_t framesInFlight,
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUFence>& fence)
 	{
+		auto logicalDevice = getLogicalDevice();
+		if (fence)
+		{
+			logicalDevice->blockForFences(1u, &fence.get());
+			if (m_frameIx >= framesInFlight) CommonAPI::dropRetiredSwapchainResources(m_qRetiredSwapchainResources, m_frameIx - framesInFlight);
+		}
+		else
+			fence = logicalDevice->createFence(static_cast<nbl::video::IGPUFence::E_CREATE_FLAGS>(0));
 	}
```

Comment:
due to implicit dependency on `m_frameIx` make it a method of `CResizableSwapchainUtils`